lib/cn/csched_sp3.c

```diff
@@ -1948,124 +2151,163 @@ sp3_schedule(struct sp3 *sp)
+            if (qfull(sp, qnum))
                 break;
 
-            case jtype_leaf_scatter:
-                /* Implements:
-                 *   - Leaf node scatter rule
-                 */
-                if (sp->thresh.lscatter_pct == 100)
+            /* Service RBT_LI_IDLE red-black tree.
+             * Implements:
+             *   - Idle node query-shape rule
+             */
+            if (sp->thresh.llen_idlec > 0) {
+                thresh = (UINT32_MAX - (jclock_ns >> 32)) << 32;
+
+                job = sp3_check_rb_tree(sp, RBT_LI_IDLE, thresh, wtype_node_idle, qnum);
+            }
+            break;
+
+        case jtype_leaf_garbage:
+            qnum = SP3_QNUM_LGARB;
+            if (qfull(sp, qnum)) {
+                qnum = SP3_QNUM_SHARED;
+                if (qfull(sp, qnum))
                     break;
-                qi = sp->qinfo + SP3_QNUM_LSCAT;
-                if (qfull(qi) && shared_full)
+            }
+
+            /* Service RBT_L_GARB red-black tree.
+             * Implements:
+             *   - Leaf node space amp rule
+             * Notes:
+             *   - Don't check for garbage unless ucomp is active
+             *     or if in samp_reduce mode and leaf percent is
+             *     somewhat caught up (ie, current leaf pct
+             *     (lpct_targ) is within 90% of rparam setting
+             *     (rp_leaf_pct)).
+             *   - When checking for garbage, if leaf percent is
+             *     behind, then bump up threshold so we don't waste
+             *     write amp by compacting nodes with a small
+             *     amount of garbage (we'd rather wait for
+             *     leaf_pct to catch up).
+             */
+            if (sp->samp_reduce && (100 * sp->lpct_targ > 90 * rp_leaf_pct)) {
+                thresh = (sp->lpct_targ < rp_leaf_pct ? 10ul : 0ul) << 32;
+
+                job = sp3_check_rb_tree(sp, RBT_L_GARB, thresh, wtype_leaf_garbage, qnum);
+            }
+            break;
+
+        case jtype_leaf_size:
+            qnum = SP3_QNUM_LSIZE;
+            if (qfull(sp, qnum)) {
+                qnum = SP3_QNUM_SHARED;
+                if (qfull(sp, qnum))
                     break;
-                job = sp3_check_rb_tree(sp, RBT_L_SCAT, SP3_LSCAT_THRESH_MIN, wtype_leaf_scatter);
+            }
+
+            /* Service RBT_L_PCAP red-black tree.
+             * - Handles big leaf nodes with or with out garbage.
+             * Implements:
+             *   - Leaf node size rule
+             */
+            job = sp3_check_rb_tree(sp, RBT_L_PCAP, 0, wtype_leaf_size, qnum);
+            break;
+
+        case jtype_leaf_scatter:
+            qnum = SP3_QNUM_SHARED;
+            if (qfull(sp, qnum))
                 break;
+
+            /* Implements:
+             *   - Leaf node scatter rule
+             */
+            if (sp->thresh.lscatter_pct < 100) {
+                thresh = (UINT32_MAX - (jclock_ns >> 32)) << 32;
```

Comment:
I was wondering whether there will be an issue as SP3_LSCAT_THRESH_MIN is not accounted for in the threshold calculation but I see that you only add nodes that are above SP3_LSCAT_THRESH_MIN to the rbtree. Nice!

Reply:
Thanks!  I tried to move all the simple and static conditions from the sp3_work() functions into sp3_dirty_node() to avoid hammering on the tree lock.  But things like run-length logic was too complicated and so we still have to call things like leaf-length to figure out whether or not there is work to be done.  But now if there is no work to be done we wont call it again until the node changes, and that was what drove the necessity for moving the timer logic out of the sp3_work() functions...

Reply:
FWIW, before this patch kvt with it's 32 kvs' would hammer on the tree lock 8000+ times per second just polling for work.  With this patch it's down to a handful!